File: kits/js/kit.js
const fs = require('fs');
const readline = require('readline');
const SEEKER = 2;
const HIDER = 3;
const DIRECTION = {
  NORTH: 0,
  NORTHEAST: 1,
  EAST: 2,
  SOUTHEAST: 3,
  SOUTH: 4,
  SOUTHWEST: 5,
  WEST: 6,
  NORTHWEST: 7,
  STILL: 8
}
const ALL_DIRECTIONS = [0, 1, 2, 3, 4, 5, 6, 7, 8];


// Create parser and use ',' as the delimiter between commands being sent by the `Match` and `MatchEngine`
const Parser = require('./parser');
const parse = new Parser(',');

class Unit {
  constructor(id, x, y, distance = undefined) {
    this.id = id;
    this.x = x;
    this.y = y;
    this.distance = distance;
  }
  move(dir) {
    return `${this.id}_${dir}`;
  }
}

/**
 * Agent for sequential `Designs`
 */
class Agent {
  _setup() {

    // Prepare to read input
    const rl = readline.createInterface({
      input: process.stdin,
      output: null,
    });

    let buffer = [];
    let currentResolve;
    const makePromise = function() {
      return new Promise((resolve) => {
        currentResolve = resolve;
      });
    };
    // on each line, push line to buffer
    rl.on('line', (line) => {
      buffer.push(line);
      currentResolve();
      currentPromise = makePromise();
    });

    // The current promise for retrieving the next line
    let currentPromise = makePromise();
    
    // with await, we pause process until there is input
    const getLine = async () => {
      return new Promise(async (resolve) => {
        while (buffer.length === 0) {
          // pause while buffer is empty, continue if new line read
          await currentPromise;
        }
        // once buffer is not empty, resolve the most recent line in stdin, and remove it
        resolve(parse(buffer.shift()));
      });
    };
    this.getLine = getLine;
  }

  constructor() {
    this._setup(); // DO NOT REMOVE
  }

  /**
   * Store unit ifnrmation
   */
  async _storeUnitInformation() {
    let unitIDsAndCoordinates = (await this.getLine()).nextStrArr();
    let units = [];
    unitIDsAndCoordinates.forEach((info) => {
      info = info.split("_");
      let unit = new Unit(parseInt(info[0]), parseInt(info[1]), parseInt(info[2]), parseInt(info[3]));
      units.push(unit);
    });
    this.units = units;

    // reset opposing units
    this.opposingUnits = [];

    let otherUnitIDsAndCoordinatesLine = (await this.getLine());
    if (otherUnitIDsAndCoordinatesLine.contents.length) {
      let otherUnitIDsAndCoordinates = otherUnitIDsAndCoordinatesLine.nextStrArr();
      let opposingUnits = [];
      otherUnitIDsAndCoordinates.forEach((info) => {
        info = info.split("_");
        let unit = new Unit(parseInt(info[0]), parseInt(info[1]), parseInt(info[2]));
        opposingUnits.push(unit);
      });
      this.opposingUnits = opposingUnits;
    }

  }

  /**
   * Initialize Agent for the `Match`
   * User should edit this according to their `Design`
   */
  async initialize() {
    
    // get agent ID
    let meta = (await this.getLine()).nextIntArr();
    this.round = 0;
    this.id = meta[0];
    this.team = meta[1]; // equals SEEKER=3 of HIDER=2

    // get unit information
    await this._storeUnitInformation();

    // get map input
    let mapInfo = (await this.getLine()).nextIntArr();
    let width = mapInfo[0];
    let height = mapInfo[1];
    this.map = [];
    for (let i = 0; i < height; i++) {
      let line = (await this.getLine()).nextIntArr();
      this.map.push(line);
    }

    // add unit ids onto the map
    this.units.forEach((unit) => {
      this.map[unit.y][unit.x] = unit.id;
    });
    this.opposingUnits.forEach((unit) => {
      this.map[unit.y][unit.x] = unit.id;
    });
    
  }
  /**
   * Updates agent's own known state of `Match`
   */
  async update() {

    this.round++;
    // reset map unit ids 
    this.units.forEach((unit) => {
      this.map[unit.y][unit.x] = 0;
    });
    this.opposingUnits.forEach((unit) => {
      this.map[unit.y][unit.x] = 0;
    });
    // get unit information
    await this._storeUnitInformation();

    // add unit ids onto the map
    this.units.forEach((unit) => {
      this.map[unit.y][unit.x] = unit.id;
    });
    this.opposingUnits.forEach((unit) => {
      this.map[unit.y][unit.x] = unit.id;
    });

  }

  move(unitID, direction) {
    return `${unitID}_${direction}`;
  }

  /**
   * End a turn
   */
  endTurn() {
    console.log('D_FINISH');
  }
}

/**
 * Applies a direction to x,y and returns an object { x: number, y: number } that 
 * represents the coordinates after going in that direction by one step.
 * @param {number} x 
 * @param {number} y 
 * @param {number} dir 
 */
function applyDirection(x, y, dir) {
  let newx = x;
  let newy = y;
  switch(dir) {
    case DIRECTION.NORTH:
      newy -=1;
      break;
    case DIRECTION.NORTHEAST:
      newy -=1;
      newx +=1;
      break;
    case DIRECTION.EAST:
      newx += 1;
      break;
    case DIRECTION.SOUTHEAST:
      newx += 1;
      newy += 1;
      break;
    case DIRECTION.SOUTH:
      newy += 1;
      break;
    case DIRECTION.SOUTHWEST:
      newy += 1;
      newx -= 1;
      break;
    case DIRECTION.WEST:
      newx -= 1;
      break;
    case DIRECTION.NORTHWEST:
      newx -= 1;
      newy -= 1;
      break;
    case DIRECTION.STILL:
      break;
  }
  return { x: newx, y: newy};
}

module.exports = { Agent, DIRECTION, ALL_DIRECTIONS, SEEKER, HIDER, applyDirection };

File: kits/python/kit.py
import sys
from enum import Enum

def apply_direction(x, y, dir):
    newx = x
    newy = y
    if (dir == 0):
        newy -= 1
    elif (dir == 1):
        newy -= 1
        newx += 1
    elif (dir == 2):
        newx += 1
    elif (dir ==3):
        newx += 1
        newy += 1
    elif (dir == 4):
        newy += 1
    elif (dir == 5):
        newy += 1
        newx -= 1
    elif (dir == 6):
        newx -= 1
    elif (dir == 7):
        newx -= 1
        newy -= 1
    elif (dir == 8):
        pass
    
    return (newx, newy)

def read_input():
    """
    Reads input from stdin
    """
    try:
        return input()
    except EOFError as eof:
        raise SystemExit(eof)
      
class Team(Enum):
    SEEKER = 2
    HIDER = 3

class Direction(Enum):
    NORTH = 0
    NORTHEAST = 1
    EAST = 2
    SOUTHEAST = 3
    SOUTH = 4
    SOUTHWEST = 5
    WEST = 6
    NORTHWEST = 7
    STILL = 8

class Unit:
    def __init__(self, id, x, y, dist):
        self.id = id
        self.x = x
        self.y = y
        self.distance = dist

    def move(self, dir: int) -> str:
        return "%d_%d" % (self.id, dir)


class Agent:
    round_num = 0
    """
    Constructor for a new agent
    User should edit this according to their `Design`
    """
    def __init__(self):
        pass

    """
    Initialize Agent for the `Match`
    User should edit this according to their `Design`
    """
    def initialize(self):
        meta = read_input().split(",")
        self.id = int(meta[0])
        self.team = Team(int(meta[1]))

        self._store_unit_info()
        

        [width, height] = [int(i) for i in (read_input().split(","))]
        self.map = []
        for i in range(height):
            line = read_input().split(",")
            parsedList = []
            for j in range(len(line)):
                if line[j] != '':
                    parsedList.append(int(line[j]))

            self.map.append(parsedList)

        self.round_num = 0

        self._update_map_with_ids()

    def _reset_map(self):

        for _, unit in enumerate(self.units):
            self.map[unit.y][unit.x] = 0

        for _, unit in enumerate(self.opposingUnits):
            self.map[unit.y][unit.x] = 0

    def _update_map_with_ids(self):

         # add unit ids to map
        for _, unit in enumerate(self.units):
            self.map[unit.y][unit.x] = unit.id
        
        # add unit ids to map
        for _, unit in enumerate(self.opposingUnits):
            self.map[unit.y][unit.x] = unit.id

    def _store_unit_info(self):
        units_and_coords = read_input().split(",")

        self.units = []
        for _, val in enumerate(units_and_coords):
            if (val != ""):
                [id, x, y, dist] = [int(k) for k in val.split("_")]
                self.units.append(Unit(id, x, y, dist))

        units_and_coords = read_input().split(",")
        
        self.opposingUnits = []
        for _, value in enumerate(units_and_coords):
            if (value != ""):
                [id, x, y] = [int(k) for k in value.split("_")]
                self.opposingUnits.append(Unit(id, x, y, -1))

    """
    Updates Agent's own known state of `Match`
    User should edit this according to their `Design
    """
    def update(self):
        self.round_num += 1
        self._reset_map()
        self._store_unit_info()
        self._update_map_with_ids()

    """
    End a turn
    """
    def end_turn(self):
        print('D_FINISH')
        
